torch/utils/_cxx_pytree.py

```diff
@@ -182,8 +181,19 @@ def register_pytree_node(
             )
         )
     """
+    from ._pytree import _register_pytree_node
+
+    _register_pytree_node(
```

Comment:
Another place where we might diverge are the default pytree nodes. What are optree's default pytree nodes?

python_pytree's are tuple, list, dict, namedtuple, and some other data structures that are spread across the codebase...

Reply:
Currently, `cxx_pytree` out-of-box support:

- [`tuple`](https://docs.python.org/3/library/stdtypes.html#tuple)
- [`list`](https://docs.python.org/3/library/stdtypes.html#list)
- [`dict`](https://docs.python.org/3/library/stdtypes.html#dict)
- [`collections.namedtuple`](https://docs.python.org/3/library/collections.html#collections.namedtuple) and its subclasses
- [`collections.OrderedDict`](https://docs.python.org/3/library/collections.html#collections.OrderedDict)
- [`collections.defaultdict`](https://docs.python.org/3/library/collections.html#collections.defaultdict)
- [`collections.deque`](https://docs.python.org/3/library/collections.html#collections.deque)
- [`PyStructSequence`](https://docs.python.org/3/c-api/tuple.html#struct-sequence-objects) types created by C API [`PyStructSequence_NewType`](https://docs.python.org/3/c-api/tuple.html#c.PyStructSequence_NewType)

Here are some major differences between cxx_pytree and python_pytree as of now:

1. (easy) cxx_pytree registers `defaultdict` and `deque` as node types while python_pytree does not yet.
2. (easy) cxx_pytree treats a sub-sub-class of a `namedtuple` type as node type, while python_pytree only allows tuple type directly inherited from `namedtuple` as node type.
3. (middle) cxx_pytree out-of-box treats all `structseq` types as node type (e.g., types in `torch.return_types`), while pytree_pytree needs to manually register them. See also:

    - #75982

4. (middle) the APIs provided by `cxx_pytree.TreeSpec` and `python_pytree.TreeSpec` are different as I mentioned in https://github.com/pytorch/pytorch/pull/110395#discussion_r1344242006.
5. (middle) the key ordering for `dict` (as well as `defaultdict`). Currently, we use insertion order, but we did not handle this well:

```python
In [1]: import torch.utils._pytree as python_pytree

In [2]: python_pytree.tree_structure({'a': 1, 'b': 2})
Out[2]: 
TreeSpec(dict, ['a', 'b'], [*,
  *])

In [3]: python_pytree.tree_structure({'b': 2, 'a': 1})
Out[3]: 
TreeSpec(dict, ['b', 'a'], [*,
  *])

In [4]: python_pytree.tree_structure({'a': 1, 'b': 2}) == python_pytree.tree_structure({'b': 2, 'a': 1})
Out[4]: False

In [5]: {'a': 1, 'b': 2} == {'b': 2, 'a': 1}
Out[5]: True
```